File: simulation/core/emission.js
/**
 * Emission Schedule Module
 * Handles token emission calculations with smooth decay
 */

import CONFIG from '../config.js';

/**
 * Calculate emission rate at a given time
 * Em(t) = Em0 * exp(-t / τ)
 * 
 * @param {number} daysSinceGenesis - Days since network genesis
 * @returns {number} Tokens emitted per epoch
 */
export function calculateEmission(daysSinceGenesis) {
    const { initialRate, decayTimeConstant } = CONFIG.emission;
    return initialRate * Math.exp(-daysSinceGenesis / decayTimeConstant);
}

/**
 * Calculate total tokens emitted from genesis to a given day
 * Integral of Em(t) from 0 to t
 * = Em0 * τ * (1 - exp(-t / τ))
 * 
 * @param {number} daysSinceGenesis - Days since network genesis
 * @returns {number} Total tokens emitted
 */
export function calculateTotalEmitted(daysSinceGenesis) {
    const { initialRate, decayTimeConstant } = CONFIG.emission;
    return initialRate * decayTimeConstant * (1 - Math.exp(-daysSinceGenesis / decayTimeConstant));
}

/**
 * Calculate remaining tokens to ever be emitted
 * Total supply = Em0 * τ
 * Remaining = Total - Emitted
 * 
 * @param {number} daysSinceGenesis - Days since network genesis
 * @returns {number} Remaining tokens to be emitted
 */
export function calculateRemainingSupply(daysSinceGenesis) {
    const { initialRate, decayTimeConstant } = CONFIG.emission;
    const totalSupply = initialRate * decayTimeConstant;
    return totalSupply - calculateTotalEmitted(daysSinceGenesis);
}

/**
 * Calculate the asymptotic total supply
 * As t → ∞, total emitted → Em0 * τ
 * 
 * @returns {number} Maximum total supply
 */
export function calculateMaxSupply() {
    const { initialRate, decayTimeConstant } = CONFIG.emission;
    return initialRate * decayTimeConstant;
}

/**
 * Get emission trajectory for visualization
 * 
 * @param {number} maxDays - Maximum days to calculate
 * @returns {Array<{day: number, emission: number, totalEmitted: number, percentEmitted: number}>}
 */
export function getEmissionTrajectory(maxDays) {
    const trajectory = [];
    const maxSupply = calculateMaxSupply();

    for (let day = 0; day <= maxDays; day++) {
        const emission = calculateEmission(day);
        const totalEmitted = calculateTotalEmitted(day);
        const percentEmitted = (totalEmitted / maxSupply) * 100;

        trajectory.push({ day, emission, totalEmitted, percentEmitted });
    }

    return trajectory;
}

/**
 * Calculate days until a certain percentage of supply is emitted
 * 
 * @param {number} targetPercent - Target percentage (0-100)
 * @returns {number} Days to reach target
 */
export function daysToEmitPercent(targetPercent) {
    const { decayTimeConstant } = CONFIG.emission;
    // percentEmitted = 100 * (1 - exp(-t / τ))
    // targetPercent/100 = 1 - exp(-t / τ)
    // exp(-t / τ) = 1 - targetPercent/100
    // t = -τ * ln(1 - targetPercent/100)
    return -decayTimeConstant * Math.log(1 - targetPercent / 100);
}

/**
 * Compare with Bitcoin-style halving for reference
 * 
 * @param {number} maxDays - Days to compare
 * @returns {Object} Comparison data
 */
export function compareWithHalving(maxDays) {
    const halvingInterval = 365 * 4; // ~4 years like Bitcoin
    const initialBlockReward = CONFIG.emission.initialRate;

    const smooth = [];
    const halving = [];

    for (let day = 0; day <= maxDays; day++) {
        const halvingCount = Math.floor(day / halvingInterval);
        const halvingEmission = initialBlockReward / Math.pow(2, halvingCount);

        smooth.push({ day, emission: calculateEmission(day) });
        halving.push({ day, emission: halvingEmission });
    }

    return { smooth, halving };
}

export default {
    calculateEmission,
    calculateTotalEmitted,
    calculateRemainingSupply,
    calculateMaxSupply,
    getEmissionTrajectory,
    daysToEmitPercent,
    compareWithHalving,
};

File: simulation_py/emission.py
"""
Emission Schedule Module
Handles token emission calculations with smooth decay
"""

import math
from config import CONFIG


def calculate_emission(days_since_genesis: float) -> float:
    """
    Calculate emission rate at a given time
    Em(t) = Em0 * exp(-t / τ)
    
    Args:
        days_since_genesis: Days since network genesis
        
    Returns:
        Tokens emitted per epoch
    """
    initial_rate = CONFIG["emission"]["initial_rate"]
    decay_constant = CONFIG["emission"]["decay_time_constant"]
    return initial_rate * math.exp(-days_since_genesis / decay_constant)


def calculate_total_emitted(days_since_genesis: float) -> float:
    """
    Calculate total tokens emitted from genesis to a given day
    Integral of Em(t) from 0 to t = Em0 * τ * (1 - exp(-t / τ))
    
    Args:
        days_since_genesis: Days since network genesis
        
    Returns:
        Total tokens emitted
    """
    initial_rate = CONFIG["emission"]["initial_rate"]
    decay_constant = CONFIG["emission"]["decay_time_constant"]
    return initial_rate * decay_constant * (1 - math.exp(-days_since_genesis / decay_constant))


def calculate_remaining_supply(days_since_genesis: float) -> float:
    """
    Calculate remaining tokens to ever be emitted
    
    Args:
        days_since_genesis: Days since network genesis
        
    Returns:
        Remaining tokens to be emitted
    """
    return calculate_max_supply() - calculate_total_emitted(days_since_genesis)


def calculate_max_supply() -> float:
    """
    Calculate the asymptotic total supply
    As t → ∞, total emitted → Em0 * τ
    
    Returns:
        Maximum total supply
    """
    initial_rate = CONFIG["emission"]["initial_rate"]
    decay_constant = CONFIG["emission"]["decay_time_constant"]
    return initial_rate * decay_constant


def get_emission_trajectory(max_days: int) -> list:
    """
    Get emission trajectory for visualization
    
    Args:
        max_days: Maximum days to calculate
        
    Returns:
        List of emission data points
    """
    trajectory = []
    max_supply = calculate_max_supply()
    
    for day in range(max_days + 1):
        emission = calculate_emission(day)
        total_emitted = calculate_total_emitted(day)
        percent_emitted = (total_emitted / max_supply) * 100
        
        trajectory.append({
            "day": day,
            "emission": emission,
            "total_emitted": total_emitted,
            "percent_emitted": percent_emitted
        })
    
    return trajectory


def days_to_emit_percent(target_percent: float) -> float:
    """
    Calculate days until a certain percentage of supply is emitted
    
    Args:
        target_percent: Target percentage (0-100)
        
    Returns:
        Days to reach target
    """
    decay_constant = CONFIG["emission"]["decay_time_constant"]
    return -decay_constant * math.log(1 - target_percent / 100)


def compare_with_halving(max_days: int) -> dict:
    """
    Compare with Bitcoin-style halving for reference
    
    Args:
        max_days: Days to compare
        
    Returns:
        Comparison data
    """
    halving_interval = 365 * 4  # ~4 years like Bitcoin
    initial_rate = CONFIG["emission"]["initial_rate"]
    
    smooth = []
    halving = []
    
    for day in range(max_days + 1):
        halving_count = day // halving_interval
        halving_emission = initial_rate / (2 ** halving_count)
        
        smooth.append({"day": day, "emission": calculate_emission(day)})
        halving.append({"day": day, "emission": halving_emission})
    
    return {"smooth": smooth, "halving": halving}
